source/develop/release-management/features/node/node-next-persistence.html.md

```diff
@@ -0,0 +1,90 @@
+
+This feature presents a mechanism by which packages installed through yum/dnf can be saved and automatically re-applied when the OS is updated.
+
+### Owner
+* Name: Ryan Barry
+* Email: rbarry@redhat.com
+
+### Current Status
+* Status: Design
+* Last updated date: 20th Sep 2016
+
+### Benefit to oVirt
+
+This feature will greatly improve the user experience of Node by maintaining a consistent userland and system configuration after upgrades. This allows for Node customizations to "stick".
+
+### Dependencies
+None! We simply need to add an additional yum plugin to imgbased.
+
+## Detailed Description
+
+Node NG performs a number of actions when updates are applied.
+
+* A new layer is added
+* The new squashfs is extracted onto that layer
+* `/etc`/ and `/root` are rsynced to the new layer
+* UIDs and GIDs are synchronized across packages, in case a drift occurred inbetween images for UIDs/GIDs which are not fixed
+* A new bootloader entry is added and set as the default
+
+The existing osupdater logic can be extended (as the new root filesystem is already mounted) in order to install a package (or packages) into the new layer before it is booted for the first time.
+
+A reliable mechanism must be written in order to capture packages as they are installed, however.
+
+[yum.PackageObject](http://yum.baseurl.org/api/yum-3.2.26/yum.packages.PackageObject-class.html) is able to immediately provide the location of a cached RPM (after downloading or from localinstall), and this is available as part of yum plugin hooks.
+
+It _may_ be possible to [extend RPM](http://www.rpm.org/wiki/DevelDocs/Plugins), but some research must be done in order to discover whether RPM's plugin interface is as informative as yum.PackageObject. It is possible that RPM's plugin mechanism only triggers during the execution of the RPM specfile, and cannot be used to pin down the location of the actual package. This also provides the additional burden of maintaining a small codebase in C, though that may be an acceptable tradeoff.
+
+## Implementation
+
+### yum plugin
+
+yum provides a [variety of places to add hooks](http://yum.baseurl.org/wiki/WritingYumPlugins), and actually somewhat more than what's listed:
+
+    args
+    clean
+    close
+    compare_providers
+    config
+    exclude
+    historybegin
+    historyend
+    init
+    postconfig
+    postdownload
+    postreposetup
+    postresolve
+    posttrans
+    postverifytrans
+    predownload
+    prelistenabledrepos
+    prereposetup
+    preresolve
+    pretrans
+    preverifytrans
+    verify_package
+
```

Comment:
The complete list is not elsewhere, hence the list here. There is a link to the list on "Writing Yum Plugins", but it's out of date.